File: ios/AtlasCore/Sources/AtlasCore/Participant/Participant.swift
import Foundation

/// Participant & profile — everyone and everything is ONE participant class. Swift parity with
/// `backend/atlas/participant.py` (Python reference of record).
///
/// A participant is a typed entity (individual / nonprofit / for-profit) operating as a persona
/// handle, with an OPT-IN public profile on the ANONYMOUS -> PUBLIC -> VERIFIED spectrum.
/// Credential authorities are NOT privileged: any participant can issue a signed Attestation about
/// a subject; VERIFICATION-NOT-AUTHORITY means trust binds to the authority KEY the consumer
/// chooses — never a spoofable display name. Atlas anoints no root.

public enum EntityClass: String, Codable, CaseIterable, Sendable {
    // Three top-level kinds: individual (verified human), organization (nonprofit/for-profit — a
    // registered collective legal entity), agent (an AI/autonomous actor that is never a root — see
    // AgentDelegation). Parity with backend/atlas/marketplace.py.
    case individual, nonprofit, forProfit = "for_profit", agent

    /// Organizations = the registered collective-entity classes (fiscal status is the sub-type).
    public var isOrganization: Bool { self == .nonprofit || self == .forProfit }
    /// Only a human or an organization can ROOT authority for an agent — never another agent.
    public var canBePrincipal: Bool { self != .agent }
}

public enum ProfileVisibility: String, Codable, Sendable { case anonymous, publicProfile = "public" }

public enum ParticipantError: Error { case wrongSubject, badAttestation, needsDisplayName }

public struct Attestation {
    public let authorityName: String          // display label ONLY — trust binds to the key
    public let authority: HybridSign.PublicKey
    public let subject: Data                  // subject persona handle
    public let claim: String
    public let epoch: Int
    public var sig: Data = Data()

    public func body() -> Data {
        Primitives.H(Data("atlas/attestation/v1".utf8), Data(authorityName.utf8), subject,
                     Data(claim.utf8), Data(String(epoch).utf8))
    }
}

public func issueAttestation(authority: HybridSign.Keypair, authorityName: String, subject: Data,
                             claim: String, epoch: Int = 0) throws -> Attestation {
    var a = Attestation(authorityName: authorityName, authority: authority.publicKey,
                        subject: subject, claim: claim, epoch: epoch)
    a.sig = try HybridSign.sign(authority, a.body())
    return a
}

public func verifyAttestation(_ a: Attestation) -> Bool {
    HybridSign.verify(a.authority, a.body(), a.sig)
}

/// A participant's self-service profile: anonymous by default; opt into public; "verified" is
/// DERIVED from holding valid authority attestations.
public final class ParticipantProfile {
    public let handle: Data
    public let entityClass: EntityClass
    public let publicKey: HybridSign.PublicKey
    public private(set) var displayName = ""
    public private(set) var bio = ""
    public private(set) var links: [String] = []
    public private(set) var visibility = ProfileVisibility.anonymous
    public private(set) var attestations: [Attestation] = []

    public init(handle: Data, publicKey: HybridSign.PublicKey, entityClass: EntityClass) {
        self.handle = handle; self.publicKey = publicKey; self.entityClass = entityClass
    }

    public func goPublic(displayName: String, bio: String = "", links: [String] = []) throws {
        guard !displayName.isEmpty else { throw ParticipantError.needsDisplayName }
        visibility = .publicProfile
        self.displayName = displayName; self.bio = bio; self.links = links
    }

    public func goAnonymous() {
        visibility = .anonymous; displayName = ""; bio = ""; links = []
    }

    /// Accept an attestation ABOUT this participant — forged or misdirected ones are rejected.
    public func hold(_ a: Attestation) throws {
        guard a.subject == handle else { throw ParticipantError.wrongSubject }
        guard verifyAttestation(a) else { throw ParticipantError.badAttestation }
        attestations.append(a)
    }

    /// Display: the authority NAMES that validly attest this participant (labels, not trust).
    public func verifiedBy() -> [String] {
        attestations.filter { $0.subject == handle && verifyAttestation($0) }.map { $0.authorityName }
    }

    public var isVerified: Bool { visibility == .publicProfile && !verifiedBy().isEmpty }
}

/// A CONSUMER's check: a valid attestation for `claim` from an authority KEY the consumer trusts.
public func presents(_ profile: ParticipantProfile, claim: String,
                     trustedAuthorityKeys: Set<Data>) -> Bool {
    profile.attestations.contains { a in
        a.subject == profile.handle && a.claim == claim && verifyAttestation(a)
            && trustedAuthorityKeys.contains(a.authority.encode())
    }
}

File: backend/atlas/participant.py
"""Participant & profile — everyone and everything is ONE participant class.

A participant is any typed entity (INDIVIDUAL / NONPROFIT / FOR_PROFIT) operating as a persona handle.
It has an OPT-IN public profile on the ANONYMOUS -> PUBLIC -> VERIFIED spectrum (self-service — anyone
makes their own), may HOLD attestations from credential authorities ("verified by X"), and — if it is
itself an authority — may ATTEST others.

Credential authorities (a medical board, a standards body, a government office, an age-verifier) are
NOT a privileged subsystem: they are just participants that issue signed attestations, discoverable and
reviewable like anyone. VERIFICATION, NOT AUTHORITY — Atlas anoints no root: an attestation is only as
good as the KEY that signed it, and a CONSUMER decides which authority keys it trusts. The authority's
NAME is a mere display label (anyone can call themselves anything); trust binds to the key.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import List, Optional, Set, Tuple

from .crypto.primitives import H
from .crypto.sign import HybridSigKeypair, HybridSigPublic, sign, verify
from .marketplace import EntityClass


class Visibility(Enum):
    ANONYMOUS = "anonymous"      # no public profile — the default
    PUBLIC = "public"            # name + bio listed (VERIFIED is derived: public + >=1 valid attestation)


class ParticipantError(Exception):
    ...


@dataclass
class Attestation:
    """A signed claim by an AUTHORITY about a SUBJECT participant ("licensed-physician",
    "age-band:18+", "accredited-university"). Verified against the authority's KEY; the authority's
    NAME is only a label. The consumer decides whether it trusts that key."""
    authority_name: str
    authority: HybridSigPublic       # trust binds HERE, not to the name
    subject: bytes                   # the subject persona handle the claim is about
    claim: str
    epoch: int = 0
    sig: bytes = b""

    def body(self) -> bytes:
        return H(b"atlas/attestation/v1", self.authority_name.encode(), self.subject,
                 self.claim.encode(), str(self.epoch).encode())


def issue_attestation(authority_kp: HybridSigKeypair, *, authority_name: str, subject: bytes,
                      claim: str, epoch: int = 0) -> Attestation:
    """An authority (any participant) signs a claim about a subject handle."""
    a = Attestation(authority_name=authority_name, authority=authority_kp.public,
                    subject=subject, claim=claim, epoch=epoch)
    a.sig = sign(authority_kp, a.body())
    return a


def verify_attestation(a: Attestation) -> bool:
    """Valid iff the authority's signature over the claim body checks out. Says nothing about whether
    the authority is TRUSTWORTHY — that is the consumer's call (see `presents`)."""
    return verify(a.authority, a.body(), a.sig)


@dataclass
class Profile:
    """A participant's self-service profile. Anonymous by default; opt into public (name/bio/links);
    'verified' is DERIVED from holding valid authority attestations."""
    handle: bytes
    entity_class: EntityClass
    public: HybridSigPublic
    display_name: str = ""
    bio: str = ""
    links: Tuple[str, ...] = ()
    visibility: Visibility = Visibility.ANONYMOUS
    attestations: List[Attestation] = field(default_factory=list)
    achievements: Tuple[str, ...] = ()                 # soulbound kinds (the SBTs live in participation/soulbound)

    def go_public(self, *, display_name: str, bio: str = "", links: Tuple[str, ...] = ()) -> None:
        """Opt in to a public profile. Reversible (set back to anonymous by clearing)."""
        if not display_name:
            raise ParticipantError("a public profile needs a display name")
        self.visibility = Visibility.PUBLIC
        self.display_name = display_name
        self.bio = bio
        self.links = tuple(links)

    def go_anonymous(self) -> None:
        self.visibility = Visibility.ANONYMOUS
        self.display_name = self.bio = ""
        self.links = ()

    def hold(self, a: Attestation) -> None:
        """Accept an attestation ABOUT this participant. Rejects a forged one or one about someone
        else — you can't collect credentials that aren't yours or aren't real."""
        if a.subject != self.handle:
            raise ParticipantError("attestation is about a different subject")
        if not verify_attestation(a):
            raise ParticipantError("attestation signature invalid")
        self.attestations.append(a)

    def verified_by(self) -> Tuple[str, ...]:
        """Display: the authority NAMES that validly attest this participant (labels, not trust)."""
        return tuple(a.authority_name for a in self.attestations
                     if a.subject == self.handle and verify_attestation(a))

    @property
    def is_verified(self) -> bool:
        return self.visibility is Visibility.PUBLIC and len(self.verified_by()) > 0


def create_profile(*, handle: bytes, public: HybridSigPublic, entity_class: EntityClass) -> Profile:
    """Self-service: anyone makes their own profile for a persona handle they control (anonymous
    until they choose to go public)."""
    return Profile(handle=handle, entity_class=entity_class, public=public)


def presents(profile: Profile, *, claim: str, trusted_authority_keys: Set[bytes]) -> bool:
    """A CONSUMER's check: does this participant hold a VALID attestation for `claim` from an authority
    whose KEY the consumer trusts? Trust binds to the key set the consumer supplies — Atlas anoints no
    root, so a consumer that doesn't trust an authority rejects it even if the signature is valid."""
    return any(a.subject == profile.handle and a.claim == claim and verify_attestation(a)
               and a.authority.encode() in trusted_authority_keys
               for a in profile.attestations)
